### DISCUSSION TITLE
E-Paper 2.9" with ESP32 D1 R32

### DISCUSSION THREAD

**Discussion by N00BTellaBrot** [micropython/micropython#9191] [Libraries & Drivers]
Hello. 

I found working code on this page: https://forum.micropython.org/viewtopic.php?f=18&t=6319&hilit=epaper+rotate#p36068
But it only works with the resolution of 212 x 104 and then only in one corner of the e-paper. 
This is how it looks: 
![20220902_151937](https://user-images.githubusercontent.com/69717254/188154990-3e1aa35a-6c08-47f7-94b3-59f0a2ce102d.jpg)

Here is my code: 
```
"""
	Example for 2.13 inch black & white & red Waveshare 2.13B E-ink screen
	Run on ESP32 Waveshare driver board (software SPI)
	Adapted by Guy -- April 2019
    We used the 2.13" E-paper in landscape mode
"""

# import epaper2in9
import epaper2in9b
# import epaper2in13b
from machine import Pin, SoftSPI

# software SPI on ESP32 Waveshare driver board
sck = Pin(18)
mosi = Pin(23)
cs = Pin(5)
busy = Pin(4)
rst = Pin(16)
dc = Pin(17)
miso = Pin(12) # miso is not used but must be declared. Let's take any unused gpio: 12
spi = SoftSPI(baudrate=100000, polarity=0, phase=0, sck=sck, mosi=mosi, miso=miso)

e = epaper2in9b.EPD(spi, cs, dc, rst, busy)
e.init()

h = 296; w = 128 # e-paper 2.9"
# h = 104;  w = 212 # e-paper heigth and width. It will be used in landscape mode

buf_black        = bytearray(w * h // 8) # used by frame buffer (landscape)
buf_red          = bytearray(w * h // 8) # used by frame buffer (landscape)
buf_epaper_black = bytearray(w * h // 8) # used to display on e-paper after bytes have been
buf_epaper_red   = bytearray(w * h // 8) # moved from frame buffer to match e-paper (portrait)

import framebuf
fb_black = framebuf.FrameBuffer(buf_black, w, h, framebuf.MONO_VLSB)
fb_red   = framebuf.FrameBuffer(buf_red,   w, h, framebuf.MONO_VLSB)
black_red = 0 # will be black on buf_black, red on buf_red
white     = 1

#clear red & black screens, write in black on top left and in red bootom right
fb_black.fill(white)
fb_red.fill(white)
fb_black.text('Hello world!', 0, 0,black_red)
fb_red.text('Hello world!', 110, 90,black_red)



# Let's draw rectangles, one black one red
fb_black.fill_rect(10, 40, 75, 10, black_red)
fb_red.fill_rect(10, 60, 75, 10, black_red)

# Move frame buffer bytes to e-paper buffer to match e-paper bytes oranisation.
# That is landscape mode to portrait mode. (for red and black buffers)
x=0; y=-1; n=0; m=0
for i in range(0, 13):
    for j in range(0, 212):
        m = (n-x)+(n-y)*12
        buf_epaper_black[m] = buf_black[n]
        buf_epaper_red[m] = buf_red[n]
        n +=1
    x = n+i+1
    y = n-1

# Now let's write text in portrait mode
fb_portrait_black   = framebuf.FrameBuffer(buf_epaper_black, h, w, framebuf.MONO_HLSB)
fb_portrait_red     = framebuf.FrameBuffer(buf_epaper_red, h, w, framebuf.MONO_HLSB)
fb_portrait_black.text('Micropython', 0, 0,black_red)
fb_portrait_red.text('Micropython', 15, 200,black_red)

# We can use EPD class to draw circle(not available in framebuf class)
# but only after we have moved bytes from framebuffers to e-paper buffers
e.draw_filled_circle(buf_epaper_black, 50, 110, 30, 1)
e.draw_filled_circle(buf_epaper_black, 50, 110, 20, 0)
e.draw_filled_circle(buf_epaper_red, 50, 150, 30, 1)
e.draw_filled_circle(buf_epaper_red, 50, 150, 20, 0)

print('Sending to display')
e.display_frame(buf_epaper_black, buf_epaper_red )
print('Done!.......')
e.sleep()  # recommended by manufacturer to avoid damage to display
print('E-paper sleeping!...')
# also recommended by manufacturer: min. 180s between 2 refresh screen

#print('Test if sleeping: Nothing should be dispayed')
#fb_black.fill(white)
#fb_red.fill(white)
#e.display_frame(buf_black, None)
print('END')
```

This is the result of my code: 
![20220902_151721](https://user-images.githubusercontent.com/69717254/188166695-ae7c1e87-51f9-434c-98fb-bab9de4d5670.jpg)

Does anyone have a solution?

**Comment by jimmo** [comment]
You need to use the driver for the correct size and type of display 

See https://github.com/waveshare/Pico_ePaper_Code and https://github.com/mcauser/micropython-waveshare-epaper

**Reply by N00BTellaBrot** [reply]
Yes I know.
But it doesn't work.

**Reply by jimmo** [reply]
You didn't mention that in your original question.

Can you be more specific about "doesn't work"? What have you tried exactly?

**Reply by N00BTellaBrot** [reply]
I got the code from this site https://forum.micropython.org/viewtopic.php?f=18&t=6319&hilit=epaper+rotate#p36068 and swapped the imports from import epaper2in13b to import epaper2in9b. Then I changed the line e = epaper2in13b.EPD(spi, cs, dc, rst, busy) to e = epaper2in9b.EPD(spi, cs, dc, rst, busy). I didn't do more.

**Reply by jimmo** [reply]
Have you tried the one from the official repository https://github.com/waveshare/Pico_ePaper_Code that I linked to?

I don't know if you have the "B" or "C" model of the 2.9" display. So you either want

https://github.com/waveshare/Pico_ePaper_Code/blob/main/python/Pico_ePaper-2.9-B.py
or
https://github.com/waveshare/Pico_ePaper_Code/blob/main/python/Pico_ePaper-2.9-C.py

**Reply by N00BTellaBrot** [reply]
Thanks for your answer. It works out. 

But do you know how to refresh the display only partially? I don't want to refresh the entire display every time.

**Reply by mcauser** [reply]
My set of drivers are about 4 years old now and no longer maintained.
The Waveshare repo has the most up to date drivers.
Mine were a fork of their original Raspberry Pi drivers, with small syntax changes to support MicroPython's SPI.
I only have 2 of their displays, but I figured I'd fork them all as they were all very similar.

Fast forward 4 years, and Waveshare have MicroPython support, a lot more examples and code for many newer displays.

Re partial refresh: check the Waveshare wiki. Not every display supports partial refresh.
If the display supports it, they'll have a set window method where you say which are you want to partial refresh, then you send the bytes for just that area.

**Reply by N00BTellaBrot** [reply]
Thank you for your answer. 

I have a last question: how do I rotate the display from portrait to landscape

I use this code: https://github.com/waveshare/Pico_ePaper_Code/blob/main/python/Pico_ePaper-2.9-C.py

I swapped the two variables EDP_HEIGHT and EDP_WIDTH, but then this is what comes out: 
![20220908_111436](https://user-images.githubusercontent.com/69717254/189084668-b9e81e56-25c0-4e0d-b616-acac04b9f18a.jpg)

**Comment by Goal-Young** [comment]
how to make the epaper display chinese
